Invoices example (#17755)
cloudflare/cloudflare-docs: src/content/docs/workflows/examples/send-invoices.mdx

---
type: example
summary: Send invoice when shopping cart is checked out and paid for
tags:
  - Workflows
  - D1
  - Email Routing
pcx_content_type: configuration
title: Pay cart and send invoice
sidebar:
  order: 3
description: Send invoice when shopping cart is checked out and paid for

---

import { TabItem, Tabs } from "~/components"

In this example, we implement a Workflow for an e-commerce website that is triggered every time a shopping cart is created.

> Once a workflow instance is triggered, it starts polling a [D1](/d1) database for the cart ID until it's been checked out. Once the shopping cart is checked out, we proceed to process the payment with an external provider doing a fetch POST. Finally, assuming everything goes well, we try to send an email using [Email Workers](/email-routing/email-workers/) with the invoice to the customer.

Reviewer: ```suggestion
Once a Workflow instance is triggered, it starts polling a [D1](/d1) database for the cart ID until it has been checked out. Once the shopping cart is checked out, we proceed to process the payment with an external provider doing a fetch POST. Finally, assuming everything goes well, we try to send an email using [Email Workers](/email-routing/email-workers/) with the invoice to the customer.
```

> As you can see, Workflows handles all the different service responses and failures; it will retry D1 until the cart is checked out, retry the payment processor if it fails for some reason, and retry sending the email with the invoice if it can't. The developer doesn't have to care about any of that logic, and the workflow can run for hours, handling all the possible conditions until it is completed.

Reviewer: ```suggestion
As you can see, Workflow handles all the different service responses and failures; it will retry D1 until the cart is checked out, retry the payment processor if it fails for some reason, and retry sending the email with the invoice if it cannot. The developer does not have to care about any of that logic, and the workflow can run for hours, handling all the possible conditions until it is completed.
```

> This is obviously a simplified example of processing a shopping cart. We'd assume more steps and additional logic in a real-life scenario, but hopefully, it gives you a good idea of what you can do with Workflows.

Reviewer: ```suggestion
This is a simplified example of processing a shopping cart. We would assume more steps and additional logic in a real-life scenario, but this example gives you a good idea of what you can do with Workflows.
```

    let url = new URL(req.url);

    let id = new URL(req.url).searchParams.get("instanceId");

    // Get the status of an existing instance, if provided
    if (id) {
      let instance = await env.CART_WORKFLOW.get(id);
      return Response.json({
        status: await instance.status(),
      });
    }

    if (url.pathname.startsWith("/new")) {
      let instance = await env.CART_WORKFLOW.create({
        params: {
          cartId: "123"
        },
      });
      return Response.json({
        id: instance.id,
        details: await instance.status(),
      });
    }

    return new Response(
      `<html><body><a href="/new">new instance</a> or add ?instanceId=...</body></html>`,
      {
        headers: {
          "content-type": "text/html;charset=UTF-8",
        },
      },
    );
  },
};
```

Here's a minimal package.json:

```json
{
    "devDependencies": {
        "@cloudflare/workers-types": "^4.20241022.0",
        "wrangler": "^3.83.0"
    },
    "dependencies": {
        "mimetext": "^3.0.24"
    }
}
```

And finally wrangler.toml:

```toml
name = "cart-invoices"
main = "src/index.ts"
compatibility_date = "2024-10-22"

> compatibility_flags = ["experimental", "nodejs_compat" ]

Reviewer 2: ```suggestion
compatibility_flags = ["nodejs_compat" ]
```
